File: 2020/4.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "aoc.h"

static const char *fields[8] =
    { "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid", "cid" };

static int isvalid(const char *key, const char *val)
{
    int d, i;
    long ld;
    unsigned int u;
    char s[SNBUF];

    s[0] = '\0';
    if (strncmp(key, "byr", 3) == 0) {
        if (sscanf(val, "%d%n", &d, &i) == EOF || i != 4 || d < 1920
            || d > 2002)
            return 0;
    } else if (strncmp(key, "iyr", 3) == 0) {
        if (sscanf(val, "%d%n", &d, &i) == EOF || i != 4 || d < 2010
            || d > 2020)
            return 0;
    } else if (strncmp(key, "eyr", 3) == 0) {
        if (sscanf(val, "%d%n", &d, &i) == EOF || i != 4 || d < 2020
            || d > 2030)
            return 0;
    } else if (strncmp(key, "hgt", 3) == 0) {
        if (sscanf(val, "%d%s", &d, s) == EOF
            || !((strcmp(s, "cm") == 0 && d >= 150 && d <= 193)
                 || (strcmp(s, "in") == 0 && d >= 59 && d <= 76)))
            return 0;
    } else if (strncmp(key, "hcl", 3) == 0) {
        if (sscanf(val, "#%x%n", &u, &i) == EOF || i != 7 || u > 0xFFFFFF)
            return 0;
    } else if (strncmp(key, "ecl", 3) == 0) {
        if (sscanf(val, "%s%n", s, &i) == EOF || i > 3
            || strncmp(s, "amb", 3)
            * strncmp(s, "blu", 3)
            * strncmp(s, "brn", 3)
            * strncmp(s, "gry", 3)
            * strncmp(s, "grn", 3)
            * strncmp(s, "hzl", 3)
            * strncmp(s, "oth", 3) != 0)
            return 0;
    } else if (strncmp(key, "pid", 3) == 0) {
        if (sscanf(val, "%ld%n", &ld, &i) == EOF || i != 9)
            return 0;
    }
    return 1;
}

static void parse(char *str, int *pr, int *va)
{
    char key[4], val[SNBUF], s[SNBUF + 4], *v;
    int i;

    hinit();
    key[3] = '\0';
    i = 0;
    do {
        /* key-value pairs are delimited by spaces or newlines */
        if (*str == ' ' || *str == '\n') {
            s[i] = '\0';
            if (sscanf(s, "%3c:%s", key, val) != EOF) {
                hstrinsert(key, val);
            }
            i = 0;
        } else {
            s[i] = *str;
            i++;
        }
        str++;
    } while (i < SNBUF + 4 && *str != '\0');

    /* field[7] is optional (cid) */
    *pr = *va = 1;
    for (i = 0; i < 7; i++) {
        v = hstrlookup(fields[i]);
        *pr &= (v != NULL);
        *va &= v ? isvalid(fields[i], v) : 0;
    }
    hfree(free);
}

int main(void)
{
    int i, c, p, na, nb, x, y;
    char passport[1024];

    na = nb = p = i = 0;
    do {
        c = getchar();
        /* passport entries are delimited by blank lines */
        if ((c == '\n' && p == '\n') || c == EOF) {
            passport[i] = '\0';
            parse(passport, &x, &y);
            na += x;
            nb += y;
            i = 0;
        } else {
            passport[i] = c;
            i++;
        }
        p = c;
    } while (i < 1023 && c != EOF);

    printf("A) %d passports with all fields present\n", na);
    printf("B) %d passports with all fields present and valid\n", nb);

    return 0;
}

File: 2020/zig/src/4.zig
const std = @import("std");
const eql = std.mem.eql;
const indexOf = std.mem.indexOf;
const parseUnsigned = std.fmt.parseUnsigned;

const fields = [_][]const u8{ "byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid" };
const ecls = [_][]const u8{ "amb", "blu", "brn", "gry", "grn", "hzl", "oth" };

fn isValid(key: []const u8, val: []const u8) bool {
    if (eql(u8, key, "byr")) {
        const n = parseUnsigned(u32, val, 10) catch return false;
        if (n >= 1920 and n <= 2002) {
            return true;
        }
    } else if (eql(u8, key, "iyr")) {
        const n = parseUnsigned(u32, val, 10) catch return false;
        if (n >= 2010 and n <= 2020) {
            return true;
        }
    } else if (eql(u8, key, "eyr")) {
        const n = parseUnsigned(u32, val, 10) catch return false;
        if (n >= 2020 and n <= 2030) {
            return true;
        }
    } else if (eql(u8, key, "hgt")) {
        if (indexOf(u8, val, "cm")) |i| {
            const n = parseUnsigned(u32, val[0..i], 10) catch return false;
            if (n >= 150 and n <= 193) {
                return true;
            }
        }
        if (indexOf(u8, val, "in")) |i| {
            const n = parseUnsigned(u32, val[0..i], 10) catch return false;
            if (n >= 59 and n <= 76) {
                return true;
            }
        }
    } else if (eql(u8, key, "hcl")) {
        if (val.len == 7) {
            const n = parseUnsigned(u32, val[1..], 16) catch return false;
            if (n <= 0xFFFFFF) {
                return true;
            }
        }
    } else if (eql(u8, key, "ecl")) {
        for (ecls) |ecl| {
            if (eql(u8, val, ecl)) {
                return true;
            }
        }
    } else if (eql(u8, key, "pid")) {
        if (val.len == 9) {
            const n = parseUnsigned(u32, val, 10) catch return false;
            return true;
        }
    }

    return false;
}

fn parse(allocator: *std.mem.Allocator, str: []u8, na: *u32, nb: *u32) !void {
    var map = std.BufMap.init(allocator);
    defer map.deinit();

    var pairs = std.mem.tokenize(str, " ");
    while (pairs.next()) |substr| {
        try map.put(substr[0..3], substr[4..]);
    }

    var valid = true;
    for (fields) |key| {
        if (map.get(key)) |val| {
            valid = valid and isValid(key, val);
        } else {
            return;
        }
    }

    na.* += 1;
    if (valid) {
        nb.* += 1;
    }
}

pub fn main() anyerror!void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = &arena.allocator;

    var p: u8 = undefined;
    var na: u32 = 0;
    var nb: u32 = 0;
    var passport = std.ArrayList(u8).init(allocator);
    defer passport.deinit();

    var stdin = std.io.getStdIn().reader();
    var in = std.io.bufferedReader(stdin).reader();
    while (in.readByte() catch null) |c| {
        // Passport entries delimited by blank lines
        if (c == '\n') {
            if (p == '\n') {
                try parse(allocator, passport.items, &na, &nb);
                passport.shrinkRetainingCapacity(0);
            } else {
                try passport.append(' ');
            }
        } else {
            try passport.append(c);
        }
        p = c;
    }
    try parse(allocator, passport.items, &na, &nb); // final password before EOF

    std.debug.print("A) {d} passports with all fields present\n", .{na});
    std.debug.print("B) {d} passports with all fields present and valid\n", .{nb});
}
